### ISSUE TITLE
clang-14 causes eBPF optimization to fail for unrolled loops

### ISSUE THREAD

**Issue by rafaeldtinoco** [aquasecurity/tracee#2579]
## Description

clang-14 causes eBPF optimization to fail for unrolled loops. Looks like clang-14 as a bug causing either bad optimization messages OR too many instructions generated by the optimizations made.

## Details (Technical)

```
        -O2 -g \
        -march=bpf -mcpu=v2 \
        -c ./pkg/ebpf/c/tracee.bpf.c \
        -o dist/tracee.bpf.core.o

./pkg/ebpf/c/tracee.bpf.c:2487:5: warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]

    for (i = 0; i < n; prefix++, str++, i++) {
    ^

./pkg/ebpf/c/tracee.bpf.c:2487:5: warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
./pkg/ebpf/c/tracee.bpf.c:2487:5: warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
```

The following code causes it:

```C
static __inline int has_prefix(char *prefix, char *str, int n)
{
    int i;
#pragma unroll
    for (i = 0; i < n; prefix++, str++, i++) {
        if (!*prefix)
            return 1;
        if (*prefix != *str) {
            return 0;
        }
    }

    // prefix is too long
    return 0;
}
```

And one way to "mitigate" the compiler warning is to replace this inline function with the following patch:

```diff
    ebpf: turn has_string to macro using compound expr
    
    Currently newer LLVM versions might fail to optimize has_string() inline
    function with:
    
    ./pkg/ebpf/c/tracee.bpf.c:2487:5: warning: loop not unrolled: the
    optimizer was unable to perform the requested transformation; the
    transformation might be disabled or specified as part of an unsupported
    transformation ordering [-Wpass-failed=transform-warning]
    
        for (i = 0; i < n; prefix++, str++, i++) {
    
    It looks like it won't recognize macro argument "n" by being a constant.
    
    By replacing has_string() inline function with a macro compound
    statement, it is possible to mitigate that issue and make eBPF binary
    properly optimized.

diff --git a/pkg/ebpf/c/tracee.bpf.c b/pkg/ebpf/c/tracee.bpf.c
index 214a8dae..07f840a5 100644
--- a/pkg/ebpf/c/tracee.bpf.c
+++ b/pkg/ebpf/c/tracee.bpf.c
@@ -2480,21 +2480,20 @@ static __always_inline int events_perf_submit(program_data_t *p, u32 id, long re
 
 // INTERNAL: STRINGS -------------------------------------------------------------------------------
 
-static __inline int has_prefix(char *prefix, char *str, int n)
-{
-    int i;
-#pragma unroll
-    for (i = 0; i < n; prefix++, str++, i++) {
-        if (!*prefix)
-            return 1;
-        if (*prefix != *str) {
-            return 0;
-        }
-    }
-
-    // prefix is too long
-    return 0;
-}
+#define has_prefix(p, s, n)                                                                     \
+    ({                                                                                          \
+        int rc = 0;                                                                             \
+        char *pre = p, *str = s;                                                                \
+        _Pragma("unroll") for (int z = 0; z < n; pre++, str++, z++)                             \
+        {                                                                                       \
+            if (!*pre)                                                                          \
+                rc = 1;                                                                         \
+            if (*pre != *str) {                                                                 \
+                rc = 0;                                                                         \
+            }                                                                                   \
+        }                                                                                       \
+        rc;                                                                                     \
+    })
```

But when done, and compiled, it generates stack limit (eBPF == 512 bytes) error messages:

```text
      -march=bpf -mcpu=v2 \
        -c ./pkg/ebpf/c/tracee.bpf.c \
        -o dist/tracee.bpf.core.o
./pkg/ebpf/c/tracee.bpf.c:5921:16: error: Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.

int BPF_KPROBE(trace_ret_vfs_write_tail)
               ^
./pkg/ebpf/c/tracee.bpf.c:5921:16: note: could not determine the original source location for ./pkg/ebpf/c/tracee.bpf.c:0:0
./pkg/ebpf/c/tracee.bpf.c:5921:16: note: could not determine the original source location for ./pkg/ebpf/c/tracee.bpf.c:0:0
./pkg/ebpf/c/tracee.bpf.c:5921:16: error: Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.

int BPF_KPROBE(trace_ret_vfs_write_tail)
               ^
./pkg/ebpf/c/tracee.bpf.c:5921:16: note: could not determine the original source location for ./pkg/ebpf/c/tracee.bpf.c:0:0
./pkg/ebpf/c/tracee.bpf.c:5921:16: note: could not determine the original source location for ./pkg/ebpf/c/tracee.bpf.c:0:0
./pkg/ebpf/c/tracee.bpf.c:5921:16: error: Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.

int BPF_KPROBE(trace_ret_vfs_write_tail)
               ^
./pkg/ebpf/c/tracee.bpf.c:5921:16: note: could not determine the original source location for ./pkg/ebpf/c/tracee.bpf.c:0:0
./pkg/ebpf/c/tracee.bpf.c:5849:13: error: Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.

        if (has_prefix(
            ^
./pkg/ebpf/c/tracee.bpf.c:5849:13: error: Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.
```

**Comment by geyslan** [comment]
@geyslan (me) keep track.

**Comment by yanivagman** [comment]
Shouldn't this have been closed by #2581?

**Comment by rafaeldtinoco** [comment]
Yes it should have been.
